### ISSUE TITLE
Open Source CLI Client

### ISSUE THREAD

**Issue by rflandau** [gravwell/gravwell#1138]
## Overview

[Link to the current repo](https://github.com/rflandau/gwcli).

This project is to create an open-source, CLI client usable via an interactive TUI or scriptable shell commands.

### Background

The current CLI is unsupported and could do with a makeover. While it has relative feature parity with the web interface, it is unpleasant to use and built on an unstable foundation. On top of this, it is currently closed-source.

Rather than attempt to overhaul and upgrade the current CLI, this project seeks to wholesale replace it.

## Milestones

- [x] 1. Invoking the CLI spawns an interactive, shell-like prompt
- [x] 2. CLI builds a nested command tree with actions as leaves and both local and global flags
- [x] 3. Actions can be invoked interactively or via the user's shell
- [x] 4. Querying and result viewing
- [x] 5. Secure login
- [ ] 6. Actions are built on a boilerplate for easy extension
- [ ] 7. Feature parity with old CLI
- [ ] 8. OPTIONAL: Further Improvements

### Path to Production

gwcli has a stable, extensible framework and a small library of genericized tools for rapidly expanding functionality. The majority of the work these past 12 weeks has been on creating a foundation upon which it is easy to maintain and upgrade Gravwell's CLI client. This foundation is flexible, fast, and simple to modify or extend. Its interactive and non-interactive interfaces are tighlty coupled for unified usage. The primary functionality, querying, is fully implemented and provides a featureful interface with many of the amenities of the web GUI.

For gwcli to be production-ready, it needs to be brought up to feature parity with the current cli client (Milestone 7). This should be primarily supported by the scaffold package, which provides genericized actions for rapidly bolting on new functionality. Milestone 6 can be completed once the edit scaffold is in place. 

Milestone 8 is extra, nice-to-have features, such as query syntax highlighting and *dynamic* tab auto-completion (rather than the semi-static currently available).

### Milestone 1

**Completed**

gwcli has a shell-like interface for interactive use, complete with interactive action invocation, contextual help, tree navigation and presence, command history, and tab auto-completion.

### Milestone 2

**Completed**

gwcli is a [Cobra](cobra.dev) command tree that is self-building, easily extensible, and highly malleable.

A set of global flags is defined on the root node. These are available to every command. Each action can also have local-only flags defined. These flags are usable in interactive and non-interactive mode (see Milestone 3).

### Milestone 3

All elements, actions, commands, et cetera of gwcli can be used interactively, invoked by the user from within the gwcli shell, or non-interactively, called directly from the user's parent shell (ex: Bash, Zsh, Powershell[?](#milestone-8)).

gwcli is intelligent enough to invoke interactivity if it needs more information from the user or if the user invokes it bare. The `--script` flag can be given to disable all interactivity (good for using gwcli in Perl, Bash, Python, et cetera).

### Milestone 4

Querying is its own action with a multiline query editor and a separate, extensible side menu for modifiers (currently just hosts query time range (duration)).

Results can be viewed and interacted with via Datascope, which provides support for viewing the results, scheduling the query, and downloading the data.

Querying can also be done entirely non-interactively. This will cause results to be output to stdout unless an output file is given.

### Milestone 5

gwcli will intelligently ask for credentials, so users do not have to put sensitive information directly into their shell. It will create a token on successful login so a user does not have to enter their credentials on every invocation.

In --script mode, gwcli will fail out if a token is not present and the user did not provide another form of identification (via flags).

### Milestone 6

The scaffold package provides simple functions for rapidly adding functionality related to one of the scaffolded action types. For instance, all list commands are built via the scaffoldlist.NewListAction() subroutine, providing them a unified interface and simple creation.

More advanced actions, like query, are still built by hand.

The scaffolds:

- [x] create
- [x] delete
- [x] list
- [x] basic
- [ ] edit

Edit, for editting existing entities (macros, kits, extractors, scheduled queries, etc) is the last common action without a scaffold.

### Milestone 7

The old CLI has significantly more feature-rich than gwcli. However, with the completion of the scaffolds in Milestone 6 and building on the back of Mother, reaching feature parity should be relatively trivial, albeit time consuming.

Current features in gwcli (old cli equivalent in parentheses, if applicable):

- [x] querying (`search`)
    - [ ] reattaching to existing queries to display them via Datascope (`attach`)
    - [ ] backgrounding queries
    - [x] result downloading (`download`) 
- scheduled searching (`schedule_search`)
    - [x] create
    - [x] delete
    - [ ] edit
    - [x] list
    - [ ] create via script
    - [ ] clear (error message or state)
    - [ ] toggle backfill
    - [ ] set timeframe offset
- saved searches and the search library
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- macros (`macro`)
    - [x] create
    - [x] delete
    - [ ] edit
    - [x] list
- alerts (`alerts`)
    - [ ] create
    - [ ] delete
    - [ ] enable/disable
    - [ ] edit
    - [ ] list
- instance state
    - IMO, most of these should be folded into a handful of basic actions under a "status" nav
    - [ ] indexer states (`state`)
    - [ ] indexer descriptions (`desc`)
    - [ ] indexer storage (`storage`)
    - [ ] system metrics (`system`)
    - [ ] index sizes (`indexes`)
    - [ ] ingester state (`ingesters`)
- self user management 
    - [ ] active sessions (`session`)
    - [ ] JWT token (`current session`) (is this necessary?)
    - [ ] view current notification for your user (`notifications`)
    - [ ] admin mode and admin actions (`admin`)
        - In the current CLI this is implemented as a submenu of admin-only commands. IMO, this should be implemented into gwcli as a toggle held in the `connection` singleton (which is already *somewhat* tracked by the gravwell Client library's AdminMode). Many calls against the Client library change their results based on the status of admin mode, so we may as well utilize this by providing an action to toggle the current state's admin status. This can be done via a basic action, likely under the 'user' nav.
    - [ ] update your information
    - [ ] change your password
    - [x] display your information (`user` > `info`)
    - [ ] edit your user info
- other user management (`admin`)
    - [ ] create a user
    - [ ] delete a user
    - [ ] impersonate a user
    - [ ] other features of the menu, as necessary
- dashboard management (`dashboards`)
    - [ ] list
    - [ ] delete
- [x] logout current session
- resources
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- kits (`kits`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [x] list
- user files (`userfiles`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- templates (`templates`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- pivots (`pivots`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- autoextractors (`extractions`)
    - [x] create
    - [ ] delete
    - [ ] edit
    - [x] list
    - [ ] create via upload
- [] direct ingestion via file (`ingest`)
- flows (`flow`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
    - [ ] create via file (`flow` > `parse`)
    - [ ] cancel running flow
    - [ ] clear (error message or state)
    - [ ] toggle backfill
- [ ] run an Anko(?) script (`script`)
- playbooks (`playbooks`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- token management (`tokens`)
    - [ ] create
    - [ ] delete
    - [ ] edit
    - [ ] list
- email (`email`)
    - [ ] delete
    - [ ] edit
- [ ] session refreshing
    - gwcli logs in on invocation. If the client idles too long, the user will lose authentication without a good way to re-authenticate (save for relaunching gwcli). This could be solved via a heartbeat goroutine to refresh the session automatically and/or a re-login command in the user nav.

### Milestone 8

This milestone is a stretch goal; gwcli should be ready for production use prior to the completion of this milestone.

On top of the more imporatnt tasks listed below, gwcli has a TODO.md file with other, non-pressing tasks that can be completed leisurely.

- [ ] Tested OS support

    By virtue of being a Go application, gwcli is likely usable on The Big Three. The `os` pkg has been used in most cases when interacting with the underlying OS. However, I do not have a good way to test gwcli on Window or Mac and primarily developed gwcli on Debian. Therefore, more work should be done to guarentee OS support.

- [ ] Chart renderers

    The chart renderer is currently unsupported. With the help of [ntchart](https://github.com/NimbleMarkets/ntcharts), however, it should be doable to get charts working in DataScope as a third mode. ntchart provides bubbles for various charts, so the main lift will be in translating the data from gravwell into a format friendly to ntchart.

- [ ] Smart weave table widths for lists

    List actions default to printing pretty tables. This is fine, but the tables do not currently size properly to terminal width (causing wrapping with looks terrible for a table), as this is slightly more advanced functionality to support (especially in non-interactive mode). Your best bet is to issue a tea.Cmd that queries for dimensions from the tty's file descriptor and returns a tea.WindowSizeMsg.
    Ex: 
    ```go
    func() tea.Msg {
		w, h, err := term.GetSize(os.Stdin.Fd())
		if err != nil {
			clilog.Writer.Errorf("Failed to fetch terminal size: %v", err)
		}
		return tea.WindowSizeMsg{Width: w, Height: h} // unless you are in Alt Buffer Mode, throw away h
	}
    ```
    Bubble Tea will submit the message once it arrives, so you can handle it in the scaffoldlist's update function.
    
    Non-interactively, you can issue the above, without the tea wrapping, as a goroutine, call the list function, wait for both to return, provide weave a table style with the given width, and then submit the list result to weave (as we are currently doing).
 
## Other Notes

The `weave` package, a library I spun off from this project, has already been merged into the [monorepo](https://github.com/gravwell/gravwell/tree/cdfc6063faf5f1b93f12b8c9de9c0f72b30f9c7a/utils/weave). It has a couple of TODOs, but these are feature improvements; I consider the package more or less complete.

**Comment by ashnwade** [comment]
Make sure the CLI works for mac users. Mike reported that `alt`+`enter` was not working to submit a query on mac.

**Comment by rflandau** [comment]
Interesting... When parsing a KeyMsg, we check the KeyMsg.Alt boolean && KeyMsg.Type == tea.KeyEnter. This not happening That would imply that Bubble Tea is not mapping the option(?) key to alt. I do not have access to a mac currently, but you can use my keyprinter code to see what information Mike's machine is sending to the terminal and how Bubble Tea is interpreting it:

```go
import tea "github.com/charmbracelet/bubbletea"

func main() {
	p := tea.NewProgram(tui{})
	if _, err := p.Run(); err != nil {
		panic(err)
	}
}

type tui struct {
}

func (t tui) Init() tea.Cmd {
	return nil
}

func (t tui) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	if keyMsg, ok := msg.(tea.KeyMsg); ok {
		if keyMsg.Type == tea.KeyEscape {
			return t, tea.Quit
		}

		return t, tea.Printf("%v | alt? %v | type: %v | paste? %v | runes: %v",
			keyMsg.String(), keyMsg.Alt, keyMsg.Type, keyMsg.Paste, keyMsg.Runes)
	}

	return t, nil
}

func (t tui) View() string {
	return "Please press a key.\nPress escape to quit."
}
```

It is also possible your prompt or terminal emulator are catching the alt key (ex: Zellij intercepts some alt key combinations for pane alternations).

**Comment by evan-gravwell** [comment]
Encountered some minor bugs while feature preview testing:
* `gwcli> status indexers stats --help` is near unreadable.
* `gwcli> queries history` shows command history, not query history (I assumed the later would show? `help` indicates such).
* `alt+enter` keybinds don't work while ssh'd into a machine on the Windows command line application.
* The carrot symbol doesn't render on the Windows command line application. I see the unknown symbol icon.

**Comment by ashnwade** [comment]
some other feature suggestions that were mentioned during the Feature Preview Meeting 2025-07-17:
- ctrl+L to clear the screen in interactive mode
- filter in place when selecting file for direct ingestion
- ability to mass chown

**Comment by rflandau** [comment]
Ctrl+L currently works, as does `clear` (to *similar* effect).

Something else to add to the list: Migrate to 1.25's JSONv2. This will primarily affect Weave.
